PROJECT TITLE "Student_placement_salary_prediction"

In [39]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

DATASET Loading

In [40]:
df = pd.read_csv(r"C:\Users\Rajesh Babu\Desktop\Machine Learning Algos\Linear_regression_student_placement_salary_prediction\student_placement_salary_elite_v2.csv")

In [41]:
df

,student_id,cgpa,branch,college_tier,python_skill,dsa_skill,ml_skill,web_dev_skill,coding_score,communication_score,aptitude_score,internships,projects,backlogs,resume_score,skill_score,placed,company_type,job_role,salary_lpa
0,S0,6.87,Civil,1,1,1,0,0,15.6,4.3,92.0,1,3,0,62.6,2,1,MNC,Software Engineer,63.55
1,S1,6.52,Civil,2,1,0,0,1,13.9,5.8,62.0,1,6,0,77.5,2,1,MNC,Data Scientist,75.17
2,S2,5.33,IT,1,1,1,1,0,9.8,8.1,66.4,0,5,1,76.0,3,1,MNC,Software Engineer,80.44
3,S3,6.04,Civil,3,1,0,1,0,39.5,9.6,83.6,0,6,0,74.3,2,1,MNC,Software Engineer,72.11
4,S4,6.78,Mechanical,2,0,1,0,1,7.5,9.9,86.3,0,3,0,66.8,2,1,Mid-size,Software Engineer,67.05
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8995,S8995,6.51,IT,2,1,0,0,1,56.6,8.2,43.2,0,2,0,50.1,2,1,MNC,Analyst,75.29
8996,S8996,7.74,IT,3,1,1,1,0,67.9,6.3,49.4,0,2,0,57.6,3,1,MNC,Analyst,69.30
8997,S8997,7.14,CSE,3,1,0,1,1,40.4,8.0,67.4,1,1,0,65.8,3,1,MNC,Data Scientist,85.58
8998,S8998,5.21,IT,1,1,0,0,1,79.9,6.3,77.2,1,6,0,78.9,2,1,Mid-size,Software Engineer,54.54


PRE PROCESSING

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   student_id           9000 non-null   object 
 1   cgpa                 9000 non-null   float64
 2   branch               9000 non-null   object 
 3   college_tier         9000 non-null   int64  
 4   python_skill         9000 non-null   int64  
 5   dsa_skill            9000 non-null   int64  
 6   ml_skill             9000 non-null   int64  
 7   web_dev_skill        9000 non-null   int64  
 8   coding_score         9000 non-null   float64
 9   communication_score  9000 non-null   float64
 10  aptitude_score       9000 non-null   float64
 11  internships          9000 non-null   int64  
 12  projects             9000 non-null   int64  
 13  backlogs             9000 non-null   int64  
 14  resume_score         9000 non-null   float64
 15  skill_score          9000 non-null   i

In [43]:
df.isnull().sum()

student_id                0
cgpa                      0
branch                    0
college_tier              0
python_skill              0
dsa_skill                 0
ml_skill                  0
web_dev_skill             0
coding_score              0
communication_score       0
aptitude_score            0
internships               0
projects                  0
backlogs                  0
resume_score              0
skill_score               0
placed                    0
company_type           1298
job_role               1298
salary_lpa                0
dtype: int64

In [44]:
df = df.drop(['student_id','branch'],axis = 1)
df

,cgpa,college_tier,python_skill,dsa_skill,ml_skill,web_dev_skill,coding_score,communication_score,aptitude_score,internships,projects,backlogs,resume_score,skill_score,placed,company_type,job_role,salary_lpa
0,6.87,1,1,1,0,0,15.6,4.3,92.0,1,3,0,62.6,2,1,MNC,Software Engineer,63.55
1,6.52,2,1,0,0,1,13.9,5.8,62.0,1,6,0,77.5,2,1,MNC,Data Scientist,75.17
2,5.33,1,1,1,1,0,9.8,8.1,66.4,0,5,1,76.0,3,1,MNC,Software Engineer,80.44
3,6.04,3,1,0,1,0,39.5,9.6,83.6,0,6,0,74.3,2,1,MNC,Software Engineer,72.11
4,6.78,2,0,1,0,1,7.5,9.9,86.3,0,3,0,66.8,2,1,Mid-size,Software Engineer,67.05
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8995,6.51,2,1,0,0,1,56.6,8.2,43.2,0,2,0,50.1,2,1,MNC,Analyst,75.29
8996,7.74,3,1,1,1,0,67.9,6.3,49.4,0,2,0,57.6,3,1,MNC,Analyst,69.30
8997,7.14,3,1,0,1,1,40.4,8.0,67.4,1,1,0,65.8,3,1,MNC,Data Scientist,85.58
8998,5.21,1,1,0,0,1,79.9,6.3,77.2,1,6,0,78.9,2,1,Mid-size,Software Engineer,54.54


In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   cgpa                 9000 non-null   float64
 1   college_tier         9000 non-null   int64  
 2   python_skill         9000 non-null   int64  
 3   dsa_skill            9000 non-null   int64  
 4   ml_skill             9000 non-null   int64  
 5   web_dev_skill        9000 non-null   int64  
 6   coding_score         9000 non-null   float64
 7   communication_score  9000 non-null   float64
 8   aptitude_score       9000 non-null   float64
 9   internships          9000 non-null   int64  
 10  projects             9000 non-null   int64  
 11  backlogs             9000 non-null   int64  
 12  resume_score         9000 non-null   float64
 13  skill_score          9000 non-null   int64  
 14  placed               9000 non-null   int64  
 15  company_type         7702 non-null   o

In [46]:
df = df[df['salary_lpa']>0]

In [47]:
df.shape

(7702, 18)

In [48]:
df_encoded = pd.get_dummies(df,columns = ['company_type','job_role'],drop_first = True)

In [49]:
df_encoded.head(10)


,cgpa,college_tier,python_skill,dsa_skill,ml_skill,web_dev_skill,coding_score,communication_score,aptitude_score,internships,...,resume_score,skill_score,placed,salary_lpa,company_type_Mid-size,company_type_Startup,company_type_Top Tech,job_role_Data Scientist,job_role_Software Engineer,job_role_Web Developer
0,6.87,1,1,1,0,0,15.6,4.3,92.0,1,...,62.6,2,1,63.55,False,False,False,False,True,False
1,6.52,2,1,0,0,1,13.9,5.8,62.0,1,...,77.5,2,1,75.17,False,False,False,True,False,False
2,5.33,1,1,1,1,0,9.8,8.1,66.4,0,...,76.0,3,1,80.44,False,False,False,False,True,False
3,6.04,3,1,0,1,0,39.5,9.6,83.6,0,...,74.3,2,1,72.11,False,False,False,False,True,False
4,6.78,2,0,1,0,1,7.5,9.9,86.3,0,...,66.8,2,1,67.05,True,False,False,False,True,False
5,9.57,1,1,0,0,0,66.9,8.0,75.5,0,...,45.2,1,1,48.75,False,True,False,True,False,False
7,8.78,1,0,0,0,0,93.0,8.8,78.0,2,...,65.6,0,1,43.16,False,True,False,True,False,False
8,6.78,3,1,0,0,0,35.3,5.8,49.9,1,...,63.6,1,1,47.32,False,True,False,False,False,True
9,7.59,2,1,0,1,1,25.2,7.0,58.1,0,...,60.8,3,1,73.21,False,False,False,False,False,False
10,8.40,1,1,0,0,0,8.1,6.2,54.5,2,...,74.9,1,1,40.65,False,True,False,False,False,True


In [50]:
df_encoded.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7702 entries, 0 to 8999
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   cgpa                        7702 non-null   float64
 1   college_tier                7702 non-null   int64  
 2   python_skill                7702 non-null   int64  
 3   dsa_skill                   7702 non-null   int64  
 4   ml_skill                    7702 non-null   int64  
 5   web_dev_skill               7702 non-null   int64  
 6   coding_score                7702 non-null   float64
 7   communication_score         7702 non-null   float64
 8   aptitude_score              7702 non-null   float64
 9   internships                 7702 non-null   int64  
 10  projects                    7702 non-null   int64  
 11  backlogs                    7702 non-null   int64  
 12  resume_score                7702 non-null   float64
 13  skill_score                 7702 non-n

MODEL TRAINING 

In [51]:
X=df_encoded.drop('salary_lpa',axis=1)

In [52]:
Y=df['salary_lpa']

In [53]:
X

,cgpa,college_tier,python_skill,dsa_skill,ml_skill,web_dev_skill,coding_score,communication_score,aptitude_score,internships,...,backlogs,resume_score,skill_score,placed,company_type_Mid-size,company_type_Startup,company_type_Top Tech,job_role_Data Scientist,job_role_Software Engineer,job_role_Web Developer
0,6.87,1,1,1,0,0,15.6,4.3,92.0,1,...,0,62.6,2,1,False,False,False,False,True,False
1,6.52,2,1,0,0,1,13.9,5.8,62.0,1,...,0,77.5,2,1,False,False,False,True,False,False
2,5.33,1,1,1,1,0,9.8,8.1,66.4,0,...,1,76.0,3,1,False,False,False,False,True,False
3,6.04,3,1,0,1,0,39.5,9.6,83.6,0,...,0,74.3,2,1,False,False,False,False,True,False
4,6.78,2,0,1,0,1,7.5,9.9,86.3,0,...,0,66.8,2,1,True,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8995,6.51,2,1,0,0,1,56.6,8.2,43.2,0,...,0,50.1,2,1,False,False,False,False,False,False
8996,7.74,3,1,1,1,0,67.9,6.3,49.4,0,...,0,57.6,3,1,False,False,False,False,False,False
8997,7.14,3,1,0,1,1,40.4,8.0,67.4,1,...,0,65.8,3,1,False,False,False,True,False,False
8998,5.21,1,1,0,0,1,79.9,6.3,77.2,1,...,0,78.9,2,1,True,False,False,False,True,False


In [54]:
Y

0       63.55
1       75.17
2       80.44
3       72.11
4       67.05
        ...  
8995    75.29
8996    69.30
8997    85.58
8998    54.54
8999    84.32
Name: salary_lpa, Length: 7702, dtype: float64

In [55]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=42)

In [56]:
X_train

,cgpa,college_tier,python_skill,dsa_skill,ml_skill,web_dev_skill,coding_score,communication_score,aptitude_score,internships,...,backlogs,resume_score,skill_score,placed,company_type_Mid-size,company_type_Startup,company_type_Top Tech,job_role_Data Scientist,job_role_Software Engineer,job_role_Web Developer
5287,5.39,2,1,1,0,0,31.2,9.4,55.8,3,...,0,91.3,2,1,True,False,False,False,False,True
5173,5.89,3,1,1,0,0,70.7,7.2,46.3,3,...,0,78.0,2,1,False,False,False,False,True,False
585,8.54,3,1,0,1,1,36.1,5.3,63.1,2,...,0,91.1,3,1,False,False,False,True,False,False
1910,5.38,2,1,1,1,1,51.8,7.7,97.0,3,...,1,120.6,4,1,True,False,False,False,False,True
7069,6.80,3,1,1,0,1,42.3,6.1,90.8,0,...,0,61.3,3,1,False,False,False,True,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6129,7.69,2,1,0,0,1,37.0,4.8,85.5,1,...,0,58.3,2,1,False,False,False,False,True,False
6324,8.44,1,1,1,0,1,98.0,8.7,47.7,1,...,0,100.4,3,1,False,False,False,False,False,False
989,6.06,3,1,1,1,0,93.1,6.7,80.2,0,...,1,69.4,3,1,True,False,False,True,False,False
8888,9.31,1,1,1,1,0,76.2,7.2,75.8,3,...,1,104.0,3,1,False,False,True,False,False,False


In [57]:
Y_test

4509    85.53
6512    66.31
7053    62.05
8855    55.26
1718    64.50
        ...  
2512    63.13
1388    45.92
8283    67.17
3436    64.93
1775    48.86
Name: salary_lpa, Length: 1541, dtype: float64

In [58]:
from sklearn.linear_model import LinearRegression


In [59]:
model=LinearRegression()

In [60]:
model.fit(X_train,Y_train)

LinearRegression()

In [61]:
Y_pred = model.predict(X_test)

In [62]:
comparision = pd.DataFrame({'Actual':Y_test,'Predicted':Y_pred})

In [63]:
comparision.head()

,Actual,Predicted
4509,85.53,84.917250
6512,66.31,61.099387
7053,62.05,59.903301
8855,55.26,60.022525
1718,64.50,60.602111


EVALUATION OF THE MODEL

In [64]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [65]:
mse = mean_squared_error(Y_test,Y_pred)
r2 = r2_score(Y_test,Y_pred)

In [66]:
print(mse)
print(r2)

43.20589198537747
0.8442779904822233


In [67]:
rmse=np.sqrt(mse)
print(rmse)
print(r2)

6.573118893293919
0.8442779904822233


CONCLUSION: The model achieved an R² score of 84.42%, indicating that it explains a significant portion of salary variation effectively.

The RMSE value of 6.57 LPA suggests that the model’s salary predictions deviate from actual salaries by approximately 6.57 LPA on average, which is reasonable considering the wide salary range present in the dataset. 